public/app/features/alerting/unified/rule-list/FilterView.tsx

```diff
@@ -146,10 +169,30 @@ function onFinished<T>(fn: () => void) {
   return tap<T>(undefined, undefined, fn);
 }
 
-function getRuleKey(ruleWithOrigin: RuleWithOrigin) {
-  const {
-    rule,
-    groupIdentifier: { rulesSource, namespace, groupName },
-  } = ruleWithOrigin;
-  return `${rulesSource.name}-${namespace.name}-${groupName}-${rule.name}-${rule.type}-${hashRule(rule)}`;
+function getRuleKey(ruleWithOrigin: RuleWithOrigin): string {
+  switch (ruleWithOrigin.origin) {
+    case 'grafana':
+      return getGrafanaRuleKey(ruleWithOrigin);
+    case 'datasource':
+      return getDataSourceRuleKey(ruleWithOrigin);
+    default:
+      const _exhaustiveCheck: never = ruleWithOrigin;
+      return _exhaustiveCheck;
+  }
```

Comment:
The `_exhaustiveCheck` was a bit confusing to me - as we only have two scenarios (`grafana` or `datasource`, could we refactor to something like this?

```tsx
function getRuleKey(ruleWithOrigin: RuleWithOrigin): string {
  if (ruleWithOrigin.origin === 'grafana') {
    return getGrafanaRuleKey(ruleWithOrigin);
  }
  return getDataSourceRuleKey(ruleWithOrigin);
}
```

(and them move the `getGrafana`/`getDataSource` methods out of the function)